# Data Wrangling and Analyzing

In this project, we'll gather, assess, and clean some Tweeter data then act on it through analysis, visualization and/or modeling.

## Table of Contents
- [Gather](#gather)
- [Access](#access)
- [Clean](#clean)
- [Analysis](#analysis)
- [Conclusion](#conclusion)

<a id='gather'></a>
## Gather

In [1]:
# Import necessary libraries
import pandas as pd
import requests
import tweepy

Load `twitter-archive-enhanced.csv` file.

In [2]:
df_tweets = pd.read_csv('twitter-archive-enhanced.csv')

Download `image_predictions.tsv` file using requests.

In [3]:
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
r = requests.get(url)

In [4]:
# Check status of response
r.status_code

200

In [5]:
# Write response to file
# with open('image_predictions.tsv', 'w+') as file:
#     file.write(r.text)    

In [6]:
df_images = pd.read_csv('image_predictions.tsv', sep='\t')

Gather tweeter information with tweepy.

In [7]:
# Connect to tweepy
# with open('keys.txt', 'r') as file:
#     api_key = file.readline().split()[2]
#     api_secret = file.readline().split()[2]
#     access_token = file.readline().split()[2]
#     access_secret = file.readline().split()[2]
    
# auth = tweepy.OAuthHandler(api_key, api_secret)
# auth.set_access_token(access_token, access_secret)

# api = tweepy.API(auth, wait_on_rate_limit = True)

In [8]:
# Download tweet data and save to tweet_json.txt
import json

# file = open('tweet_json.txt', 'w+')
# count = 0
# for tweet_id in df['tweet_id']:
#     count = count + 1
#     print('{count} / 2356, {id}'.format(count=count, id=tweet_id))
#     try:
#         file.write(json.dumps(api.get_status(id=tweet_id)._json) + '\n')
#     except:
#         print('Deleted: ', tweet_id)
#         continue
# file.close()

In [9]:
# Extract retweet count and favorite count of each tweet
likes = []
with open('tweet_json.txt', 'r') as file:
    for line in file:
        tweet = json.loads(line)
        tweet_id = tweet['id_str']
        retweet = tweet['retweet_count'] 
        favorite = tweet['favorite_count']
      
        likes.append({'tweet_id': tweet_id,
                        'retweet_count': retweet,
                        'favorite_count':favorite})
df_likes = pd.DataFrame(likes, columns = ['tweet_id', 'retweet_count', 'favorite_count'])

<a id='access'></a>
## Access

Now we have 3 files dataframes on hand, `df_tweets`, `df_images`, and `df_likes`.

In [10]:
df_tweets.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [11]:
df_tweets.sample(3)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
675  789530877013393408                    NaN                  NaN   
282  839239871831150596                    NaN                  NaN   
245  845812042753855489                    NaN                  NaN   

                     timestamp  \
675  2016-10-21 18:16:44 +0000   
282  2017-03-07 22:22:32 +0000   
245  2017-03-26 01:38:00 +0000   

                                                source  \
675  <a href="http://twitter.com/download/iphone" r...   
282  <a href="http://twitter.com/download/iphone" r...   
245  <a href="http://twitter.com/download/iphone" r...   

                                                  text  retweeted_status_id  \
675  This is Rizzy. She smiles a lot. 12/10 contagi...                  NaN   
282  This is Odie. He's big. 13/10 would attempt to...                  NaN   
245  We usually don't rate polar bears but this one...                  NaN   

     retweeted_status_user_id retweeted_status_timestamp  \
675                       NaN                        NaN   
282                       NaN                        NaN   
245                       NaN                        NaN   

                                         expanded_urls  rating_numerator  \
675  https://twitter.com/dog_rates/status/789530877...                12   
282  https://twitter.com/dog_rates/status/839239871...                13   
245  https://twitter.com/dog_rates/status/845812042...                13   

     rating_denominator   name doggo floofer pupper puppo  
675                  10  Rizzy  None    None   None  None  
282                  10   Odie  None    None   None  None  
245                  10   None  None    None   None  None

In [12]:
df_tweets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [13]:
df_tweets.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [14]:
sum(df_tweets.tweet_id.duplicated())

0

In [15]:
df_tweets.rating_numerator.sort_values()

315        0
1016       0
2335       1
2261       1
2338       1
        ... 
2074     420
188      420
189      666
313      960
979     1776
Name: rating_numerator, Length: 2356, dtype: int64

In [16]:
df_tweets.query('rating_numerator == 1776')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
979  749981277374128128                    NaN                  NaN   

                     timestamp  \
979  2016-07-04 15:00:45 +0000   

                                                source  \
979  <a href="https://about.twitter.com/products/tw...   

                                                  text  retweeted_status_id  \
979  This is Atticus. He's quite simply America af....                  NaN   

     retweeted_status_user_id retweeted_status_timestamp  \
979                       NaN                        NaN   

                                         expanded_urls  rating_numerator  \
979  https://twitter.com/dog_rates/status/749981277...              1776   

     rating_denominator     name doggo floofer pupper puppo  
979                  10  Atticus  None    None   None  None

In [17]:
df_tweets.rating_denominator.sort_values()

313       0
2335      2
516       7
1576     10
1575     10
       ... 
1635    110
1779    120
1634    130
902     150
1120    170
Name: rating_denominator, Length: 2356, dtype: int64

In [18]:
df_tweets.query('rating_denominator == 0 or rating_denominator == 2 or rating_denominator == 7')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
313   835246439529840640           8.352460e+17           26259576.0   
516   810984652412424192                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   

                      timestamp  \
313   2017-02-24 21:54:03 +0000   
516   2016-12-19 23:06:23 +0000   
2335  2015-11-16 16:11:11 +0000   

                                                 source  \
313   <a href="http://twitter.com/download/iphone" r...   
516   <a href="http://twitter.com/download/iphone" r...   
2335  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
313   @jonnysun @Lin_Manuel ok jomny I know you're e...                  NaN   
516   Meet Sam. She smiles 24/7 &amp; secretly aspir...                  NaN   
2335  This is an Albanian 3 1/2 legged  Episcopalian...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
313                        NaN                        NaN   
516                        NaN                        NaN   
2335                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
313                                                 NaN               960   
516   https://www.gofundme.com/sams-smile,https://tw...                24   
2335  https://twitter.com/dog_rates/status/666287406...                 1   

      rating_denominator  name doggo floofer pupper puppo  
313                    0  None  None    None   None  None  
516                    7   Sam  None    None   None  None  
2335                   2    an  None    None   None  None

In [19]:
df_tweets[df_tweets.expanded_urls.isnull()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
30    886267009285017600           8.862664e+17         2.281182e+09   
55    881633300179243008           8.816070e+17         4.738443e+07   
64    879674319642796034           8.795538e+17         3.105441e+09   
113   870726314365509632           8.707262e+17         1.648776e+07   
148   863427515083354112           8.634256e+17         7.759620e+07   
179   857214891891077121           8.571567e+17         1.806710e+08   
185   856330835276025856                    NaN                  NaN   
186   856288084350160898           8.562860e+17         2.792810e+08   
188   855862651834028034           8.558616e+17         1.943518e+08   
189   855860136149123072           8.558585e+17         1.361572e+07   
218   850333567704068097           8.503288e+17         2.195506e+07   
228   848213670039564288           8.482121e+17         4.196984e+09   
234   847617282490613760           8.476062e+17         4.196984e+09   
274   840698636975636481           8.406983e+17         8.405479e+17   
290   838150277551247360           8.381455e+17         2.195506e+07   
291   838085839343206401           8.380855e+17         2.894131e+09   
313   835246439529840640           8.352460e+17         2.625958e+07   
342   832088576586297345           8.320875e+17         3.058208e+07   
346   831926988323639298           8.319030e+17         2.068372e+07   
375   828361771580813312                    NaN                  NaN   
387   826598799820865537           8.265984e+17         4.196984e+09   
409   823333489516937216           8.233264e+17         1.582854e+09   
427   821153421864615936           8.211526e+17         1.132119e+08   
498   813130366689148928           8.131273e+17         4.196984e+09   
513   811647686436880384           8.116272e+17         4.196984e+09   
570   801854953262350336           8.018543e+17         1.185634e+07   
576   800859414831898624           8.008580e+17         2.918590e+08   
611   797165961484890113           7.971238e+17         2.916630e+07   
701   786051337297522688           7.727430e+17         7.305050e+17   
707   785515384317313025                    NaN                  NaN   
843   766714921925144576           7.667118e+17         4.196984e+09   
857   763956972077010945           7.638652e+17         1.584641e+07   
967   750381685133418496           7.501805e+17         4.717297e+09   
1005  747651430853525504           7.476487e+17         4.196984e+09   
1080  738891149612572673           7.384119e+17         3.589728e+08   
1295  707983188426153984           7.079801e+17         2.319108e+09   
1345  704491224099647488           7.044857e+17         2.878549e+07   
1445  696518437233913856                    NaN                  NaN   
1446  696490539101908992           6.964887e+17         4.196984e+09   
1474  693644216740769793           6.936422e+17         4.196984e+09   
1479  693582294167244802           6.935722e+17         1.198989e+09   
1497  692423280028966913           6.924173e+17         4.196984e+09   
1523  690607260360429569           6.903413e+17         4.670367e+08   
1598  686035780142297088           6.860340e+17         4.196984e+09   
1605  685681090388975616           6.855479e+17         4.196984e+09   
1618  684969860808454144           6.849598e+17         4.196984e+09   
1663  682808988178739200           6.827884e+17         4.196984e+09   
1689  681340665377193984           6.813394e+17         4.196984e+09   
1774  678023323247357953           6.780211e+17         4.196984e+09   
1819  676590572941893632           6.765883e+17         4.196984e+09   
1844  675849018447167488           6.758457e+17         4.196984e+09   
1895  674742531037511680           6.747400e+17         4.196984e+09   
1905  674606911342424069           6.744689e+17         4.196984e+09   
1914  674330906434379776           6.658147e+17         1.637468e+07   
1940  673716320723169284           6.737159e+17         

In [20]:
df_tweets[df_tweets.name == "None"]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
5     891087950875897856                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
...                  ...                    ...                  ...   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2351  666049248165822465                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
5     2017-07-29 00:08:17 +0000   
7     2017-07-28 00:22:40 +0000   
12    2017-07-25 01:55:32 +0000   
24    2017-07-18 16:08:03 +0000   
25    2017-07-18 00:07:08 +0000   
...                         ...   
2342  2015-11-16 02:38:37 +0000   
2343  2015-11-16 01:59:36 +0000   
2344  2015-11-16 01:52:02 +0000   
2351  2015-11-16 00:24:50 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
5     <a href="http://twitter.com/download/iphone" r...   
7     <a href="http://twitter.com/download/iphone" r...   
12    <a href="http://twitter.com/download/iphone" r...   
24    <a href="http://twitter.com/download/iphone" r...   
25    <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2342  <a href="http://twitter.com/download/iphone" r...   
2343  <a href="http://twitter.com/download/iphone" r...   
2344  <a href="http://twitter.com/download/iphone" r...   
2351  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
5     Here we have a majestic great white breaching ...                  NaN   
7     When you watch your owner call another dog a g...                  NaN   
12    Here's a puppo that seems to be on the fence a...                  NaN   
24    You may not have known you needed to see this ...                  NaN   
25    This... is a Jubilant Antarctic House Bear. We...                  NaN   
...                                                 ...                  ...   
2342  Here we have a well-established sunblockerspan...                  NaN   
2343  Let's hope this flight isn't Malaysian (lol). ...                  NaN   
2344  Here we have a northern speckled Rhododendron....                  NaN   
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
5                          NaN                        NaN   
7                          NaN                        NaN   
12                         NaN                        NaN   
24                         NaN                        NaN   
25                         NaN                        NaN   
...                        ...                        ...   
2342                       NaN                        NaN   
2343                       NaN                        NaN   
2344                       NaN                        NaN   
2351                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
5     https://twitter.com/dog_rates/status/891087950...                13   
7     https://twitter.com/dog_rates/status/890729181...                13   
12    https://twitter.com/dog_rates/status/889665388...                13   
24    https://twitt

In [21]:
df_images.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [22]:
df_images.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [23]:
df_images.sample(3)

tweet_id                                          jpg_url  \
1724  819952236453363712  https://pbs.twimg.com/media/C2EONHNWQAUWxkP.jpg   
721   685973236358713344  https://pbs.twimg.com/media/CYURBGoWYAAKey3.jpg   
1461  778286810187399168  https://pbs.twimg.com/media/Cs0HuUTWcAUpSE8.jpg   

      img_num                              p1   p1_conf  p1_dog  \
1724        1  American_Staffordshire_terrier  0.925505    True   
721         1                  Siberian_husky  0.450678    True   
1461        1                     Boston_bull  0.322070    True   

                             p2   p2_conf  p2_dog                 p3  \
1724  Staffordshire_bullterrier  0.036221    True  Italian_greyhound   
721                  Eskimo_dog  0.430275    True           malamute   
1461                        pug  0.229903    True             muzzle   

       p3_conf  p3_dog  
1724  0.020412    True  
721   0.118590    True  
1461  0.101420   False

In [24]:
sum(df_images.tweet_id.duplicated())

0

In [25]:
df_images.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [26]:
df_likes.head()

tweet_id  retweet_count  favorite_count
0  892420643555336193           7305           34812
1  892177421306343426           5457           30164
2  891815181378084864           3598           22706
3  891689557279858688           7490           38065
4  891327558926688256           8051           36342

In [27]:
sum(df_likes.duplicated())

0

In [28]:
df_likes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2331 entries, 0 to 2330
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   tweet_id        2331 non-null   object
 1   retweet_count   2331 non-null   int64 
 2   favorite_count  2331 non-null   int64 
dtypes: int64(2), object(1)
memory usage: 54.8+ KB


In [29]:
df_likes.describe()

retweet_count  favorite_count
count    2331.000000     2331.000000
mean     2569.824539     7269.266409
std      4346.939571    11291.352969
min         1.000000        0.000000
25%       521.500000     1258.000000
50%      1197.000000     3149.000000
75%      2975.000000     8885.500000
max     73712.000000   149541.000000

### Quality

##### `df_tweets` table
- ~~Erroneous datatypes (`tweet_id`, `in_reply_to_status_id`, `in_reply_to_user_id`, `timestamp`, `retweeted_status_id`, `retweeted_status_user_id`, `retweet_status_timestamp columns`)~~
- ~~Missing information on `in_reply_to_status_id`, `in_reply_to_user_id`, `retweeted_status_id`, `retweeted_status_user_id`, and `retweet_status_timestamp`~~
- ~~`rating_numerators` and `rating_denominators` are not matching the rating given in `text`~~
- ~~Rating numerators having very large range, some are not actual ratings (e.g. 1776 is not rating for a dog but the year of Declaration of Independence of the United States)~~
- ~~745 dogs are named "None", some are named "a"~~
- ~~`rating_numerator` and `rating_denominator` are wrong for `tweet_id` '666287406224695296', it should be 9/10 instead of 1/2~~
- ~~`rating_numerator` is wrong for `tweet_id` '883482846933004288', it should be 13.5 instead of 5~~
- ~~`rating_numerator` and `rating_denominator` are wrong for `tweet_id` '810984652412424192', 24/7 is not a rating. There is no rating for this dog~~
- ~~Some of the tweets are not original tweets~~
- ~~Some of the tweets contain videos instead of images~~

##### `df_images` table
- ~~Erroneous datatype (`tweet_id`)~~
- ~~Mix use of upper and lower cases first letters of prediction (p1, p2, p3)~~

##### `df_likes` table
- ~~Number of entries does not match number of entries of `df_tweets` - there are some deleted tweets~~
- ~~Erroneous datatype (`retweet_count`, `favorite_count`)~~

### Tidiness

##### `df_tweets` table
- ~~Four columns of dog stages~~
- ~~Retweet counts and favorite counts should be part of the `df_tweets` table~~

<a id='clean'></a>
## Clean

In [30]:
tweets_clean = df_tweets.copy()
images_clean = df_images.copy()
likes_clean = df_likes.copy()

### Quality

#### `tweets`: Some of the tweets are not original tweets

##### Define

Since we only care about original tweets with images, we are dropping all the retweets.

##### Code

In [31]:
# Make a mask of all tweets with 'in_reply_to_user_id' (meaning it's a reply), 
# 'retweeted_status_id' (meaning it's a retweet), and withough 'expanded_urls'
# (meaning it does not contain images)

mask = (tweets_clean['in_reply_to_user_id'].notnull()) | (tweets_clean['retweeted_status_id'].notnull())

In [32]:
tweets_clean.drop(tweets_clean[mask].index, axis = 0, inplace=True)

In [33]:
# Remove in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id, retweeted_status_user_id, 
# retweeted_status_timestamp because these rows have been emptied
tweets_clean.drop(columns=['in_reply_to_status_id', 'in_reply_to_user_id', 'retweeted_status_id', 
                           'retweeted_status_user_id', 'retweeted_status_timestamp'], inplace=True)

##### Test

In [34]:
tweets_clean.head()

tweet_id                  timestamp  \
0  892420643555336193  2017-08-01 16:23:56 +0000   
1  892177421306343426  2017-08-01 00:17:27 +0000   
2  891815181378084864  2017-07-31 00:18:03 +0000   
3  891689557279858688  2017-07-30 15:58:51 +0000   
4  891327558926688256  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   
1  This is Tilly. She's just checking pup on you....   
2  This is Archie. He is a rare Norwegian Pouncin...   
3  This is Darla. She commenced a snooze mid meal...   
4  This is Franklin. He would like you to stop ca...   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [35]:
tweets_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   tweet_id            2097 non-null   int64 
 1   timestamp           2097 non-null   object
 2   source              2097 non-null   object
 3   text                2097 non-null   object
 4   expanded_urls       2094 non-null   object
 5   rating_numerator    2097 non-null   int64 
 6   rating_denominator  2097 non-null   int64 
 7   name                2097 non-null   object
 8   doggo               2097 non-null   object
 9   floofer             2097 non-null   object
 10  pupper              2097 non-null   object
 11  puppo               2097 non-null   object
dtypes: int64(3), object(9)
memory usage: 213.0+ KB


#### `tweets`: Some of the tweets contain videos instead of images

##### Define

Remove all tweets without an expanded_urls or expanded_urls contains the word 'vine'

##### Code

In [36]:
# Find all the expanded_urls that are either null or contains the word "vine"
mask = (tweets_clean['expanded_urls'].isnull()) | (tweets_clean['expanded_urls'].str.contains('vine', regex=False))

In [37]:
tweets_clean.drop(tweets_clean[mask].index, axis = 0, inplace=True)

##### Test

In [38]:
tweets_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2003 entries, 0 to 2355
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   tweet_id            2003 non-null   int64 
 1   timestamp           2003 non-null   object
 2   source              2003 non-null   object
 3   text                2003 non-null   object
 4   expanded_urls       2003 non-null   object
 5   rating_numerator    2003 non-null   int64 
 6   rating_denominator  2003 non-null   int64 
 7   name                2003 non-null   object
 8   doggo               2003 non-null   object
 9   floofer             2003 non-null   object
 10  pupper              2003 non-null   object
 11  puppo               2003 non-null   object
dtypes: int64(3), object(9)
memory usage: 203.4+ KB


In [39]:
tweets_clean[tweets_clean['expanded_urls'].str.contains('vine')]

Empty DataFrame
Columns: [tweet_id, timestamp, source, text, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

In [40]:
tweets_clean[tweets_clean['expanded_urls'].isnull()]

Empty DataFrame
Columns: [tweet_id, timestamp, source, text, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

#### `tweets_clean`: Erroneous datatypes (`tweet_id`, `in_reply_to_status_id`, `in_reply_to_user_id`, `timestamp`,  `retweeted_status_id`, `retweeted_status_user_id`, `retweet_status_timestamp` columns)

##### Define

Change `tweet_id` to string, change `timestamp` to datetime

##### Code

In [41]:
tweets_clean.tweet_id = tweets_clean.tweet_id.astype(str)
tweets_clean.timestamp = pd.to_datetime(tweets_clean.timestamp)

##### Test

In [42]:
tweets_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2003 entries, 0 to 2355
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype              
---  ------              --------------  -----              
 0   tweet_id            2003 non-null   object             
 1   timestamp           2003 non-null   datetime64[ns, UTC]
 2   source              2003 non-null   object             
 3   text                2003 non-null   object             
 4   expanded_urls       2003 non-null   object             
 5   rating_numerator    2003 non-null   int64              
 6   rating_denominator  2003 non-null   int64              
 7   name                2003 non-null   object             
 8   doggo               2003 non-null   object             
 9   floofer             2003 non-null   object             
 10  pupper              2003 non-null   object             
 11  puppo               2003 non-null   object             
dtypes: datetime64[ns, UTC](1), int64(2

#### `images_clean`: Erroneous datatype (`tweet_id`)

##### Define

Change `tweet_id` to string

##### Code

In [43]:
images_clean.tweet_id = images_clean.tweet_id.astype(str)

##### Test

In [44]:
images_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   object 
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(1), object(5)
memory usage: 152.1+ KB


#### `likes_clean`: Erroneous datatype (`retweet_count`, `favorite_count`)

##### Define

Change `retweet_count` and `favorite_count` to int

##### Code

In [45]:
likes_clean.retweet_count = likes_clean.retweet_count.astype(int)
likes_clean.favorite_count = likes_clean.favorite_count.astype(int)

##### Test

In [46]:
likes_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2331 entries, 0 to 2330
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   tweet_id        2331 non-null   object
 1   retweet_count   2331 non-null   int32 
 2   favorite_count  2331 non-null   int32 
dtypes: int32(2), object(1)
memory usage: 36.5+ KB


### Tidiness

#### `tweets_clean`: Four columns of dog stages

##### Define

Merge the `doggo`, `floofer`, `pupper`, and `puppo` columns to a `dog_stages` column.


##### Code

In [47]:
# Replace 'None' and NaN with an empty string for columns 'doggo', 
# 'floofer', 'pupper' and 'puppo'

import numpy as np

tweets_clean['doggo'].replace('None', '', inplace=True)
tweets_clean['doggo'].replace(np.NaN, '', inplace=True)
tweets_clean['floofer'].replace('None', '', inplace=True)
tweets_clean['floofer'].replace(np.NaN, '', inplace=True)
tweets_clean['pupper'].replace('None', '', inplace=True)
tweets_clean['pupper'].replace(np.NaN, '', inplace=True)
tweets_clean['puppo'].replace('None', '', inplace=True)
tweets_clean['puppo'].replace(np.NaN, '', inplace=True)

In [48]:
# Extract information from tweets_clean['text'] to see if dog stage is mentioned
tweets_clean['dog_stages'] = tweets_clean['text'].str.extract('(doggo|floofer|pupper|puppo)', expand = True)

In [49]:
tweets_clean['dog_stages'].value_counts()

pupper     223
doggo       73
puppo       28
floofer      3
Name: dog_stages, dtype: int64

In [50]:
# Combine dog stages - some dogs have multiple stages
tweets_clean['dog_stages'] = tweets_clean['doggo'] + tweets_clean['floofer'] + tweets_clean['pupper'] + tweets_clean['puppo']
tweets_clean.loc[tweets_clean['dog_stages'] == 'doggopupper', 'dog_stages'] = 'doggo, pupper'
tweets_clean.loc[tweets_clean['dog_stages'] == 'doggopuppo', 'dog_stages'] = 'doggo, puppo'
tweets_clean.loc[tweets_clean['dog_stages'] == 'doggofloofer', 'dog_stages'] = 'doggo, floofer'

In [51]:
# Drop the original 'doggo','floofer','pupper','puppo' columns
tweets_clean.drop(['doggo','floofer','pupper','puppo'], axis=1, inplace = True)

##### Test

In [52]:
tweets_clean['dog_stages'].value_counts()

                  1694
pupper             204
doggo               66
puppo               22
doggo, pupper        8
floofer              7
doggo, puppo         1
doggo, floofer       1
Name: dog_stages, dtype: int64

#### `likes_clean`: Retweet counts and favorite counts should be part of the `tweets_clean` table

##### Define

Merge `likes_clean` and `tweets_clean`


##### Code

In [53]:
tweets_clean = pd.merge(tweets_clean, likes_clean, how='left', on='tweet_id')

In [54]:
# Null entries for retweet and favorite counts are due to deleted tweets
tweets_clean['retweet_count'] = tweets_clean['retweet_count'].fillna(0)
tweets_clean['favorite_count'] = tweets_clean['favorite_count'].fillna(0)

##### Test

In [55]:
tweets_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2003 entries, 0 to 2002
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype              
---  ------              --------------  -----              
 0   tweet_id            2003 non-null   object             
 1   timestamp           2003 non-null   datetime64[ns, UTC]
 2   source              2003 non-null   object             
 3   text                2003 non-null   object             
 4   expanded_urls       2003 non-null   object             
 5   rating_numerator    2003 non-null   int64              
 6   rating_denominator  2003 non-null   int64              
 7   name                2003 non-null   object             
 8   dog_stages          2003 non-null   object             
 9   retweet_count       2003 non-null   float64            
 10  favorite_count      2003 non-null   float64            
dtypes: datetime64[ns, UTC](1), float64(2), int64(2), object(6)
memory usage: 187.8+ KB


#### `tweets_clean`: `rating_numerators` and `rating_denominators` are not matching the rating given in `text`.

##### Define

Extract numbers from text and rewrite the `rating_numerator` and `rating_denominator` column.

##### Code

In [56]:
tweets_clean['rating'] = tweets_clean.text.str.extract('(\d+\.?\d*\/\d*\.?\d+)')

In [57]:
tweets_clean['rating'].unique()

array(['13/10', '12/10', '14/10', '13.5/10', '11/10', '6/10', '10/10',
       '0/10', '84/70', '24/7', '9.75/10', '5/10', '11.27/10', '3/10',
       '7/10', '8/10', '9/10', '4/10', '165/150', '1776/10', '9/11',
       '204/170', '4/20', '50/50', '99/90', '80/80', '45/50', '60/50',
       '44/40', '121/110', '7/11', '11.26/10', '2/10', '144/120', '88/80',
       '1/10', '420/10', '1/2'], dtype=object)

In [58]:
ratings = tweets_clean['rating'].str.split('/', expand = True)

In [59]:
tweets_clean['rating_numerator'] = ratings[0]

In [60]:
tweets_clean['rating_denominator'] = ratings[1]

In [61]:
tweets_clean.rating_numerator = tweets_clean.rating_numerator.astype(float)
tweets_clean.rating_denominator = tweets_clean.rating_denominator.astype(float)

In [62]:
tweets_clean.drop(columns=['rating'], inplace=True)

#### Test

In [63]:
tweets_clean.rating_numerator.sort_values()

246        0.0
1982       1.0
1985       1.0
1745       1.0
1533       1.0
         ...  
1452     144.0
685      165.0
870      204.0
1728     420.0
749     1776.0
Name: rating_numerator, Length: 2003, dtype: float64

In [64]:
tweets_clean.rating_denominator.sort_values()

1982      2.0
402       7.0
0        10.0
1342     10.0
1341     10.0
        ...  
969      90.0
1320    110.0
1452    120.0
685     150.0
870     170.0
Name: rating_denominator, Length: 2003, dtype: float64

#### `tweets_clean`: Rating numerators having very large range, some are not actual ratings (e.g. 1776 is not rating for a dog but the year of Declaration of Independence of the United States)

##### Define

Inpect outlier `rating_numerators` and deicde if deletion is needed.

##### Code

In [65]:
tweets_clean.query('rating_numerator == 0')

tweet_id                 timestamp  \
246  835152434251116546 2017-02-24 15:40:31+00:00   

                                                source  \
246  <a href="http://twitter.com/download/iphone" r...   

                                                  text  \
246  When you're so blinded by your systematic plag...   

                                         expanded_urls  rating_numerator  \
246  https://twitter.com/dog_rates/status/835152434...               0.0   

     rating_denominator  name dog_stages  retweet_count  favorite_count  
246                10.0  None                    2861.0         21618.0

In [66]:
# rating_numerator == 0 is a tweet about plagiarism, remove
tweets_clean.drop(tweets_clean[tweets_clean.rating_numerator == 0].index, axis = 0, inplace=True)

In [67]:
tweets_clean.query('rating_numerator == 1776')

tweet_id                 timestamp  \
749  749981277374128128 2016-07-04 15:00:45+00:00   

                                                source  \
749  <a href="https://about.twitter.com/products/tw...   

                                                  text  \
749  This is Atticus. He's quite simply America af....   

                                         expanded_urls  rating_numerator  \
749  https://twitter.com/dog_rates/status/749981277...            1776.0   

     rating_denominator     name dog_stages  retweet_count  favorite_count  
749                10.0  Atticus                    2361.0          4927.0

In [68]:
# 1776 is not rating for a dog but the year of Declaration of Independence of the United States
tweets_clean.drop(tweets_clean[tweets_clean.rating_numerator == 1776].index, axis = 0, inplace=True)

In [69]:
tweets_clean.query('rating_numerator == 24 and rating_denominator == 7')

tweet_id                 timestamp  \
402  810984652412424192 2016-12-19 23:06:23+00:00   

                                                source  \
402  <a href="http://twitter.com/download/iphone" r...   

                                                  text  \
402  Meet Sam. She smiles 24/7 &amp; secretly aspir...   

                                         expanded_urls  rating_numerator  \
402  https://www.gofundme.com/sams-smile,https://tw...              24.0   

     rating_denominator name dog_stages  retweet_count  favorite_count  
402                 7.0  Sam                    1387.0          5198.0

In [70]:
# 24/7 is not rating for a dog
tweets_clean.drop(tweets_clean[tweets_clean.rating_numerator == 24].index, axis = 0, inplace=True)

In [71]:
tweets_clean.query('rating_denominator == 2')

tweet_id                 timestamp  \
1982  666287406224695296 2015-11-16 16:11:11+00:00   

                                                 source  \
1982  <a href="http://twitter.com/download/iphone" r...   

                                                   text  \
1982  This is an Albanian 3 1/2 legged  Episcopalian...   

                                          expanded_urls  rating_numerator  \
1982  https://twitter.com/dog_rates/status/666287406...               1.0   

      rating_denominator name dog_stages  retweet_count  favorite_count  
1982                 2.0   an                      57.0           130.0

In [72]:
# The correct rating for this tweet is 9/10 instead of 1/2
tweets_clean.drop(tweets_clean[tweets_clean.rating_denominator == 2].index, axis = 0, inplace=True)

#### Test

In [73]:
tweets_clean.query('rating_numerator == 0 or rating_numerator == 1776 or rating_denominator == 2 or rating_denominator == 7')

Empty DataFrame
Columns: [tweet_id, timestamp, source, text, expanded_urls, rating_numerator, rating_denominator, name, dog_stages, retweet_count, favorite_count]
Index: []

In [74]:
tweets_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1999 entries, 0 to 2002
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype              
---  ------              --------------  -----              
 0   tweet_id            1999 non-null   object             
 1   timestamp           1999 non-null   datetime64[ns, UTC]
 2   source              1999 non-null   object             
 3   text                1999 non-null   object             
 4   expanded_urls       1999 non-null   object             
 5   rating_numerator    1999 non-null   float64            
 6   rating_denominator  1999 non-null   float64            
 7   name                1999 non-null   object             
 8   dog_stages          1999 non-null   object             
 9   retweet_count       1999 non-null   float64            
 10  favorite_count      1999 non-null   float64            
dtypes: datetime64[ns, UTC](1), float64(4), object(6)
memory usage: 187.4+ KB


#### `tweets`: 745 dogs are named "None", some are named "a"

##### Define

Change dog names from "None"/"a" to empty string.

##### Code

In [75]:
tweets_clean['name'].replace('None', '', inplace=True)
tweets_clean['name'].replace('a', '', inplace=True)

##### Test

In [76]:
tweets_clean['name'].value_counts()

           608
Charlie     11
Lucy        10
Oliver      10
Cooper      10
          ... 
Anna         1
Eevee        1
Vince        1
Skye         1
Eugene       1
Name: name, Length: 934, dtype: int64

In [77]:
sum(tweets_clean['name'].isnull())

0

#### `images_clean`: Mix use of upper and lower cases first letters of prediction (p1, p2, p3)

##### Define

Change every string in `p1`, `p2`, `p3` to lower case

##### Code

In [78]:
images_clean['p1'] = images_clean['p1'].str.lower()
images_clean['p2'] = images_clean['p2'].str.lower()
images_clean['p3'] = images_clean['p3'].str.lower()

##### Test

In [79]:
images_clean.p1.str.islower().value_counts()

True    2075
Name: p1, dtype: int64

In [80]:
images_clean.p2.str.islower().value_counts()

True    2075
Name: p2, dtype: int64

In [81]:
images_clean.p3.str.islower().value_counts()

True    2075
Name: p3, dtype: int64

<a id='analysis'></a>
## Analysis

<a id='conclusion'></a>
## Conclusion